File: Prosary.Tests/SevenSorrowsEngineTests.cs
using Prosary.Models;
using Prosary.Services;
using Xunit;

namespace Prosary.Tests;

/// <summary>Mirrors iOS's SevenSorrowsEngineTests.swift / Android's SevenSorrowsEngineTest.kt.</summary>
public class SevenSorrowsEngineTests
{
    private readonly PrayerEngine _engine = new(new LiturgicalCalendarService());

    private static IReadOnlyList<RosaryStep> BuildSteps(PrayerEngine engine, string? languageCode) =>
        engine.BuildSteps(new Prayer { Kind = PrayerKind.SevenSorrows, LanguageCode = languageCode ?? LanguageCatalog.DefaultSentinel });

    [Fact]
    public void BuildSteps_SevenDecadesOfSevenHailMarys()
    {
        var steps = BuildSteps(_engine, "en");
        var decadeIndices = steps.Where(s => s.DecadeIndex.HasValue).Select(s => s.DecadeIndex!.Value).Distinct();
        Assert.Equal(Enumerable.Range(0, 7), decadeIndices.OrderBy(i => i));

        for (var d = 0; d < 7; d++)
        {
            var hailMarysInDecade = steps.Count(s => s.DecadeIndex == d && s.HailMaryIndexInDecade.HasValue);
            Assert.Equal(7, hailMarysInDecade);
        }
    }

    [Fact]
    public void BuildSteps_NoStepHasAMystery()
    {
        // Seven Sorrows steps deliberately leave Mystery null (see BeadLayout's generalization)
        // — the Seven Sorrows aren't Rosary "mysteries", and unlike Franciscan Crown, none of the
        // seven reuse an existing mystery imageKey either.
        var steps = BuildSteps(_engine, "en");
        Assert.All(steps, s => Assert.Null(s.Mystery));
    }

    [Fact]
    public void BuildSteps_ThreeClosingHailMarysForOurLadysTears()
    {
        var steps = BuildSteps(_engine, "en");
        var nonDecadeHailMarys = steps.Count(s => !s.DecadeIndex.HasValue && s.Title.StartsWith("Hail Mary"));
        Assert.Equal(3, nonDecadeHailMarys);
    }

    [Fact]
    public void BuildSteps_EndsWithClosingPrayerThenClosingCross()
    {
        var steps = BuildSteps(_engine, "en");
        Assert.Equal("Our Lady of Sorrows", steps[^2].Title);
        Assert.Equal("Sign of the Cross", steps[^1].Title);
    }

    [Fact]
    public void BuildSteps_FirstSorrowIsProphecyOfSimeon()
    {
        var steps = BuildSteps(_engine, "en");
        var firstSorrow = steps.First(s => s.DecadeIndex == 0 && !s.HailMaryIndexInDecade.HasValue && s.Title != "Our Father");
        Assert.Equal("The Prophecy of Simeon", firstSorrow.Title);
        Assert.Equal("seven_sorrows_01_prophecy_of_simeon", firstSorrow.ImageOverrideKey);
        Assert.True(firstSorrow.IsScripture);
    }

    [Fact]
    public void BuildSteps_FourthSorrowHasNoScriptureCitation()
    {
        // "Mary Meets Jesus on the Way of the Cross" isn't narrated in any Gospel — a
        // traditional devotional scene, not a quoted verse — so unlike the other six, it's not
        // IsScripture.
        var steps = BuildSteps(_engine, "en");
        var fourthSorrow = steps.First(s => s.DecadeIndex == 3 && !s.HailMaryIndexInDecade.HasValue && s.Title != "Our Father");
        Assert.Equal("Mary Meets Jesus on the Way of the Cross", fourthSorrow.Title);
        Assert.False(fourthSorrow.IsScripture);
    }

    [Fact]
    public void BuildSteps_EnglishBodyContainsEnglishText()
    {
        var steps = BuildSteps(_engine, "en");
        Assert.Contains(steps, s => s.Body.Contains("Hail Mary, full of grace"));
    }

    [Fact]
    public void BuildSteps_LatinBodyContainsLatinText()
    {
        var steps = BuildSteps(_engine, "la");
        Assert.Contains(steps, s => s.Body.Contains("Ave Maria, gratia plena"));
    }
}

File: iOS/ProsaryTests/SevenSorrowsEngineTests.swift
//
//  SevenSorrowsEngineTests.swift
//  ProsaryTests
//

import XCTest
@testable import Prosary

final class SevenSorrowsEngineTests: XCTestCase {
  private let engine = PrayerEngine()

  func testSevenDecadesOfSevenHailMarys() {
    let steps = engine.buildSteps(for: Prayer(kind: .sevenSorrows, languageCode: "en"))
    let decadeIndices = Set(steps.compactMap(\.decadeIndex))
    XCTAssertEqual(decadeIndices, Set(0..<7))

    for d in 0..<7 {
      let hailMarysInDecade = steps.filter { $0.decadeIndex == d && $0.hailMaryIndexInDecade != nil }
      XCTAssertEqual(hailMarysInDecade.count, 7, "decade \(d) should have 7 Hail Marys")
    }
  }

  func testNoStepHasAMystery() {
    // Seven Sorrows steps deliberately leave `mystery` nil (see BeadModels' generalization) —
    // the Seven Sorrows aren't Rosary "mysteries", and unlike Franciscan Crown, none of the
    // seven reuse an existing mystery imageKey either.
    let steps = engine.buildSteps(for: Prayer(kind: .sevenSorrows, languageCode: "en"))
    XCTAssertTrue(steps.allSatisfy { $0.mystery == nil })
  }

  func testThreeClosingHailMarysForOurLadysTears() {
    let steps = engine.buildSteps(for: Prayer(kind: .sevenSorrows, languageCode: "en"))
    let nonDecadeHailMarys = steps.filter { $0.decadeIndex == nil && $0.title.hasPrefix("Hail Mary") }
    XCTAssertEqual(nonDecadeHailMarys.count, 3)
  }

  func testEndsWithClosingPrayerThenClosingCross() {
    let steps = engine.buildSteps(for: Prayer(kind: .sevenSorrows, languageCode: "en"))
    XCTAssertEqual(steps[steps.count - 2].title, "Our Lady of Sorrows")
    XCTAssertEqual(steps.last?.title, "Sign of the Cross")
  }

  func testFirstSorrowIsProphecyOfSimeon() {
    let steps = engine.buildSteps(for: Prayer(kind: .sevenSorrows, languageCode: "en"))
    let firstSorrow = steps.first { $0.decadeIndex == 0 && $0.hailMaryIndexInDecade == nil && $0.title != "Our Father" }
    XCTAssertEqual(firstSorrow?.title, "The Prophecy of Simeon")
    XCTAssertEqual(firstSorrow?.imageKey, "seven_sorrows_01_prophecy_of_simeon")
    XCTAssertTrue(firstSorrow?.isScripture ?? false)
  }

  func testFourthSorrowHasNoScriptureCitation() {
    // "Mary Meets Jesus on the Way of the Cross" isn't narrated in any Gospel — a traditional
    // devotional scene, not a quoted verse — so unlike the other six, it's not isScripture.
    let steps = engine.buildSteps(for: Prayer(kind: .sevenSorrows, languageCode: "en"))
    let fourthSorrow = steps.first { $0.decadeIndex == 3 && $0.hailMaryIndexInDecade == nil && $0.title != "Our Father" }
    XCTAssertEqual(fourthSorrow?.title, "Mary Meets Jesus on the Way of the Cross")
    XCTAssertFalse(fourthSorrow?.isScripture ?? true)
  }

  func testEnglishBodyContainsEnglishText() {
    let steps = engine.buildSteps(for: Prayer(kind: .sevenSorrows, languageCode: "en"))
    XCTAssertTrue(steps.contains { $0.body.contains("Hail Mary, full of grace") })
  }

  func testLatinBodyContainsLatinText() {
    let steps = engine.buildSteps(for: Prayer(kind: .sevenSorrows, languageCode: "la"))
    XCTAssertTrue(steps.contains { $0.body.contains("Ave Maria, gratia plena") })
  }
}
